### PR TITLE
feat: extract session metrics from OTEL /v1/traces spans

### PR DESCRIPTION
**Pull request by not-stbenjam**
## Summary
- Extract session metrics from OTEL `/v1/traces` spans (session_id, duration, num_turns, ttft_ms, prompt, skills, etc.) — previously only `/v1/metrics` and `/v1/logs` were parsed
- Unpin `agentic-ci` in Dockerfile so the image picks up the latest release (includes SIGTERM fix for OTEL trace flushing)
- Extract initial user prompt from `claude_code.interaction` spans (earliest by timestamp)
- Validate required fields (`session_id`, `duration_ms`, `num_turns`, `total_cost_usd`, `prompt`) — `extract_metrics.py` exits non-zero with a clear error if any are empty/zero
- Stream-log enrichment no longer overwrites trace-derived values — traces take priority
- Add `requirements-dev.txt`, `make test` runs pytest + all plugin tests, new GitHub Actions workflow

## Motivation
After migrating the payload agent prow job to `agentic-ci run --backend local` ([openshift/release#80680](https://github.com/openshift/release/pull/80680)), the `--stream-log` is no longer available. Session metrics were producing zeroed fields. This extracts those fields from OTEL traces and adds validation so silent regressions (e.g. OTEL schema changes) are caught.

### Against real CI OTEL data
Verified against 223-record `claude-otel.jsonl` from [successful rehearsal](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/80680/rehearse-80680-periodic-ci-openshift-release-main-payload-agent-analyze/2069070652526039040):

```
session_id=bf5c68e0-... model=claude-opus-4-6 cost=$6.999115
duration=871s turns=15 tools=106 subagents=2
claude_code_version=2.1.176 ttft_ms=2487 skills=3 files_written=3
```

### Multi-run prompt extraction
Ran 2 sequential `agentic-ci run` invocations with different prompts, concatenated OTEL JSONL. Confirmed `extract_metrics.py` correctly captures the **first** prompt (earliest `startTimeUnixNano`), not subsequent ones.

### Required field validation
OTEL data with only `/v1/metrics` (no traces) correctly exits non-zero:
```
ERROR: required fields are empty or zero: session_id, duration_ms, num_turns, total_cost_usd, prompt
```

### Tests
All 4 plugin test files pass (`make test`):
- `test_extract_metrics.py` — 7 tests including `test_missing_required_fields_errors`
- `test_validate.py` (autodl-json) — 10 tests
- `test_validate.py` (results-yaml) — 7 tests
- `test_sample_and_estimate.py` — 27 tests

## Test plan
- [x] E2E local test with agentic-ci 0.3.9 (SIGTERM fix confirmed)
- [x] Multi-run prompt extraction verified
- [x] Required field validation verified
- [x] All plugin tests pass
- [ ] Verify in next prow rehearsal after image rebuild

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Extracts and populates session-scoped metrics from OpenTelemetry trace payloads, including session identity, timing data, interaction counts, user prompts, and tool usage statistics
  * Validates required fields in metrics output with error reporting

* **Tests**
  * Added comprehensive test coverage for OpenTelemetry trace parsing and required field validation

* **Chores**
  * Version bumped to 0.0.6
  * Added GitHub Actions CI/CD workflow for automated testing
  * Updated development dependencies and testing infrastructure

### CODE REVIEW

**Comment by stbenjam** [comment]
/ok-to-test
/approve

(Automated: Extending self-approval rights and ok-to-test to my bot account.)

File: images/Dockerfile [hunk 1]
```diff
@@ -37,7 +37,7 @@ COPY --from=claude-latest /usr/bin/claude /usr/local/bin/claude-latest
 
 # Install Python tools and PCP bindings
 RUN python3 -m pip install --no-cache-dir \
-    agentic-ci==0.3.4 \
+    agentic-ci \
```
[the remark this answers is not in the document]
Comment (stbenjam) [reply images/Dockerfile:40]: agentic-ci is an internal project

**Comment by enxebre** [comment]
/code-review:pr suggestions :

  High — Add direct unit tests for TraceAccumulator (Skill/Agent tool spans, multi-batch accumulation, prompt ordering)

  High — Add a test for the stream-log fallback path (traces produce zero, stream-log fills in)

  Medium — Remove dead code: redundant int() cast at extract_metrics.py:~93

  Medium — Simplify int(float(str(ttft))) to int(float(ttft))

  Low — Unify attribute extraction across metrics/logs/traces (follow-up)

  Low — Extract span name magic strings to module-level constants

  Low — Use row["key"] instead of row.get("key", default) in validation where keys are contractual``

/lgtm
/hold
in case you want to address any

**Comment by not-stbenjam** [comment]
Addressed review feedback in 828b4ee:

**High priority (done):**
- Added direct unit tests for `TraceAccumulator`: Skill/Agent/Write tool spans, multi-batch accumulation across multiple `/v1/traces` payloads, and prompt ordering by earliest interaction timestamp
- Added test for stream-log fallback path (when traces produce zero for duration/ttft/num_turns, stream-log fills them in)

**Medium priority (done):**
- Removed redundant `int()` cast on already-int `start_ns` (was line ~95)
- Simplified `int(float(str(ttft)))` to `int(float(ttft))`

**Low priority (noted for follow-up):**
- Unifying attribute extraction across metrics/logs/traces
- Extracting span name magic strings to module-level constants
- Using `row["key"]` instead of `row.get("key", default)` in validation

File: plugins/prow-agent/scripts/extract_metrics.py [hunk 2]
```diff
@@ -193,12 +313,15 @@ def enrich_from_stream_log(row, stream_log_path):
...
         last_result = results[-1]
-        row["duration_ms"] = str(sum(r.get("duration_ms", 0) for r in results))
-        row["ttft_ms"] = str(results[0].get("ttft_ms", 0))
-        row["num_turns"] = str(sum(r.get("num_turns", 0) for r in results))
+        if row["duration_ms"] in ("", "0"):
+            row["duration_ms"] = str(sum(r.get("duration_ms", 0) for r in results))
+        if row["ttft_ms"] in ("", "0"):
+            row["ttft_ms"] = str(results[0].get("ttft_ms", 0))
+        if row["num_turns"] in ("", "0"):
+            row["num_turns"] = str(sum(r.get("num_turns", 0) for r in results))
```
Comment (petr-muller) [inline plugins/prow-agent/scripts/extract_metrics.py:321]: **Blocking: `files_written` and `num_subagents` were missed in the non-destructive guard refactor.**

This PR correctly added `if`-guards for `duration_ms`, `ttft_ms`, and `num_turns` (lines 316–321) so trace-derived values aren't overwritten. But the same pattern was not applied to `files_written` (line 355) and `num_subagents` (line 360) in `enrich_from_stream_log`:

```python
row["files_written"] = str(files_written_count)  # always overwrites
row["num_subagents"] = str(len(task_starts))      # always overwrites
```

When `--stream-log` is provided and the log produces zero counts (different format, incomplete capture), valid trace-derived values are silently discarded. Fix:

```python
if row["files_written"] in ("", "0"):
    row["files_written"] = str(files_written_count)
if row["num_subagents"] in ("", "0"):
    row["num_subagents"] = str(len(task_starts))
```

File: plugins/prow-agent/scripts/extract_metrics.py [hunk 3]
```diff
@@ -16,6 +16,118 @@
...
+                        if context == "interaction":
+                            self.num_turns += 1
+                            if start_ns is not None:
+                                if (self.first_interaction_start_ns is None
+                                        or start_ns < self.first_interaction_start_ns):
+                                    self.first_interaction_start_ns = start_ns
+                                    ttft = attrs.get("ttft_ms")
+                                    if ttft is not None:
+                                        self.first_ttft_ms = int(float(ttft))
+                        self.last_stop_reason = str(attrs.get("stop_reason", ""))
```
Comment (petr-muller) [inline plugins/prow-agent/scripts/extract_metrics.py]: **Should fix: `last_stop_reason` is updated by all `llm_request` spans, not just interaction ones.**

This line is outside the `if context == "interaction":` block, so it runs for every `claude_code.llm_request` span regardless of context:

```python
if context == "interaction":
    self.num_turns += 1
    ...
self.last_stop_reason = str(attrs.get("stop_reason", ""))  # ← runs for ALL llm_request spans
```

A non-interaction span (tool-use sub-call, background request) processed after the final interaction span will overwrite the meaningful stop_reason — e.g., replacing `"end_turn"` with `"tool_use"` or `""`. OTEL payloads arrive in batches and spans within a batch are not guaranteed to be in start-time order.

Fix: move the assignment inside the `if context == "interaction":` block.

File: plugins/prow-agent/scripts/extract_metrics.py [hunk 4]
```diff
@@ -193,12 +313,15 @@ def enrich_from_stream_log(row, stream_log_path):
...
+            row["ttft_ms"] = str(results[0].get("ttft_ms", 0))
+        if row["num_turns"] in ("", "0"):
+            row["num_turns"] = str(sum(r.get("num_turns", 0) for r in results))
         row["is_error"] = str(1 if last_result.get("is_error", False) else 0)
         row["terminal_reason"] = last_result.get("terminal_reason", "")
-        row["stop_reason"] = last_result.get("stop_reason", "")
+        row["stop_reason"] = row["stop_reason"] or last_result.get("stop_reason", "")
 
     seen_tool_ids = set()
     skills = []
```
Comment (petr-muller) [inline plugins/prow-agent/scripts/extract_metrics.py:327]: **Should fix: `skills_invoked` overwritten by stream-log when non-empty, losing trace-derived skills.**

Below this point (around line 354), `enrich_from_stream_log` does:

```python
if skills:
    row["skills_invoked"] = ",".join(dict.fromkeys(skills))
```

This replaces the trace-derived skill set whenever the stream-log finds any skills — inconsistent with how `duration_ms`, `ttft_ms`, and `num_turns` are handled (trace wins, stream-log fills gaps). If the two sources parse different subsets, the trace-derived list is silently dropped. Should this use the same or-fill pattern: `row["skills_invoked"] = row["skills_invoked"] or ",".join(...)`?

**Review by petr-muller — CHANGES_REQUESTED** [review]
Two confirmed bugs in the stream-log enrichment path need fixing before merge. The `TraceAccumulator` design and field-priority semantics are solid for timing fields, but the same guard logic was not applied to `files_written` and `num_subagents`, and `stop_reason` is captured from all LLM request spans regardless of context.

**Comment by petr-muller** [comment]
The above is my Claude having some opinions about the Python code, I do not consider them blocking. Unhold as needed.

/lgtm
/hold

**Comment by not-stbenjam** [comment]
Addressed all three review threads from @petr-muller in 8cebb1b:

1. **files_written & num_subagents guard** — added `if row[...] in ("", "0")` guards so trace-derived values aren't overwritten by stream-log
2. **stop_reason scoped to interaction spans** — moved `last_stop_reason` assignment inside the `if context == "interaction"` block so non-interaction llm_request spans can't overwrite it
3. **skills_invoked or-fill** — changed to `row["skills_invoked"] = row["skills_invoked"] or ...` so trace-derived skills take priority

All 11 tests pass.
